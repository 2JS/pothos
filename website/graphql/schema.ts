import searchIndex from '../util/search-index';
import {
  DocsIndex,
  DocsSection,
  SearchMatch,
  SearchMatcher,
  SearchMatchKind,
  SearchResult,
} from '../util/search-matcher';
import { builder } from './builder';

const DocsPage = builder.objectRef<DocsIndex>('DocsPage').implement({
  fields: (t) => ({
    title: t.exposeString('name'),
    description: t.exposeString('description'),
    link: t.exposeString('link'),
    md: t.exposeString('md'),
    html: t.exposeString('html'),
  }),
});

builder.enumType(SearchMatchKind, { name: 'SearchMatchKind' });

const SearchMatchRef = builder.objectRef<SearchMatch>('SearchMatch').implement({
  fields: (t) => ({
    kind: t.expose('kind', { type: SearchMatchKind }),
    section: t.expose('section', { type: DocsSectionRef, nullable: true }),
  }),
});

const SearchResultRef = builder.objectRef<SearchResult>('SearchResult').implement({
  fields: (t) => ({
    matches: t.expose('matches', { type: [SearchMatchRef] }),
    doc: t.expose('doc', { type: DocsPage, nullable: true }),
  }),
});

const DocsSectionRef = builder.objectRef<DocsSection>('DocsSection').implement({
  fields: (t) => ({
    heading: t.exposeString('heading'),
    link: t.exposeString('link'),
    md: t.exposeString('md'),
    html: t.exposeString('html'),
  }),
});

builder.queryType({
  fields: (t) => ({
    search: t.field({
      type: [SearchResultRef],
      args: {
        query: t.arg.string({
          required: true,
        }),
      },
      resolve: (_, { query }) => new SearchMatcher(query).matchDocs(searchIndex).getResults(),
    }),
  }),
});

export const schema = builder.toSchema();
