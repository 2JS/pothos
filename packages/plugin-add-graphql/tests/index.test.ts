import { lexicographicSortSchema, printSchema } from 'graphql';
import exampleSchema from './examples/simple/schema';

describe('simple objects example schema', () => {
  it('generates expected schema', () => {
    expect(printSchema(lexicographicSortSchema(exampleSchema))).toMatchSnapshot();
  });
});
