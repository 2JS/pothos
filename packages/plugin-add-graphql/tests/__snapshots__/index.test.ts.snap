// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`simple objects example schema > generates expected schema 1`] = `
"type AddedUser implements Node {
  bio: Profile
  id: ID!
  name: String
  posts: [Post!]
}

interface Node {
  id: ID!
}

type Post implements Node {
  content: String
  id: ID!
  postTitle: String!
}

type Profile implements Node {
  bio: String
  id: ID!
}

type Query {
  posts: [Post!]!
  user(id: ID!): AddedUser!
}"
`;
